Orange/tests/test_freeviz.py

```diff
@@ -0,0 +1,101 @@
+
+
+class TestFreeviz(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.iris = Table("iris")
+        cls.housing = Table("housing")
+        cls.zoo = Table("zoo")
+
+    def test_basic(self):
+        table = Table("iris")
+        table[3, 3] = np.nan
+        freeviz = FreeViz()
+        model = freeviz(table)
+        model(table)
+
+    def test_regression(self):
+        table = Table("housing")[::10]
+        freeviz = FreeViz()
+        freeviz(table)
+
+        freeviz = FreeViz(p=2)
+        freeviz(table)
+
+    def test_prepare_freeviz_data(self):
+        table = Table("iris")
+        FreeViz.prepare_freeviz_data(table)
+
+        table.X = table.X * np.nan
+        self.assertEqual(FreeViz.prepare_freeviz_data(table), (None, None, None))
+
+        table.X = None
+        FreeViz.prepare_freeviz_data(table)
+
+    @unittest.skip("Test weights is too slow.")
+    def test_weights(self):
+        table = Table("iris")
+        weights = np.random.rand(150, 1).flatten()
+        freeviz = FreeViz(weights=weights, p=3, scale=False, center=False)
+        freeviz(table)
+
+        scale = np.array([0.5, 0.4, 0.6, 0.8])
+        freeviz = FreeViz(scale=scale, center=[0.2, 0.6, 0.4, 0.2])
+        freeviz(table)
+
+    def test_raising_errors(self):
+        table = Table("iris")
+        freeviz = FreeViz(initial=(2, 4))
+        self.assertRaises(ValueError, freeviz, table)
+
+        scale = np.array([0.5, 0.4, 0.6])
+        freeviz = FreeViz(scale=scale)
+        self.assertRaises(ValueError, freeviz, table)
+
+        freeviz = FreeViz(center=[0.6, 0.4, 0.2])
+        self.assertRaises(ValueError, freeviz, table)
+
+        weights = np.random.rand(100, 1).flatten()
+        freeviz = FreeViz(weights=weights)
+        self.assertRaises(ValueError, freeviz, table)
+
+    def test_initial(self):
+        FreeViz.init_radial(1)
+        FreeViz.init_radial(2)
+        FreeViz.init_radial(3)
+        FreeViz.init_random(2, 4, 5)
+
+    def test_transform_changed_domain(self):
+        """
+        Thest where:
+        1. Open data, apply some preprocessor, splits the data into two parts,
+        use FreeViz on the first part, and then transform the second part.
+        2. Open data, split into two parts, apply the same preprocessor and
+        FreeViz only on the first part, and then transform the second part.
+
+        The transformed second part in (1) and (2) has to be the same.
+        """
+        data = Table("titanic")[::10]
+        normalize = Continuize()
+        freeviz = FreeViz(initial=FreeViz.init_radial(8), maxiter=40)
+
+        # normalize all
+        ndata = normalize(data)
+        model = freeviz(ndata[:100])
+        result_1 = model(ndata[100:])
+
+        # normalize only the "training" part
+        ndata = normalize(data[:100])
+        model = freeviz(ndata)
+        result_2 = model(data[100:])
+
+        self.assertTrue(np.allclose(result_1, result_2))
```

Comment:
Why not `np.testing.assert_almost_equal`?

@markotoplak, the test is otherwise OK, right?

Reply:
> ======================================================================
> ERROR: test_transform_changed_domain (Orange.tests.test_freeviz.TestFreeviz)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/travis/build/biolab/orange3/build/travis-test/Orange/tests/test_freeviz.py", line 101, in test_transform_changed_domain
>     np.testing.assert_almost_equal(result_1, result_2)
>   File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/numpy/testing/utils.py", line 568, in assert_almost_equal
>     if not (gisfinite(desired) and gisfinite(actual)):
> ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
>

Reply:
```python
numpy.testing.assert_allclose
```

Reply:
The test is almost OK.

To make results repeatable, init_random should be fixed so that it uses rstate = np.random.RandomState(0) if not given. Then, inititalization should be removed from the test.

The comment on the test is superfluous - the code and two in code comments say it all.